service-netty/src/main/java/info/bitrich/xchangestream/service/netty/NettyStreamingService.java

```diff
@@ -312,7 +312,7 @@ public void sendMessage(String message) {
         LOG.info("Subscribing to channel {}", channelId);
 
         return Observable.<T>create(e -> {
-            if (webSocketChannel == null || !webSocketChannel.isOpen()) {
+            if (webSocketChannel == null) { // || !webSocketChannel.isOpen()) {
```

Comment:
Could you explain the logic here?

Also, if this check should be removed, it should be removed not commented out.

Reply:
This is NettyStreamingService so I get the challenge.

| webSocketChannel == null 
Was there before and I do not challenge that logic

| || !webSocketChannel.isOpen()) {
This is unnecessary because sendMessage checks it too. 

Sorry I am just not that great at GIT (CVS was my tech horizon - I kid of course, i got on with subversion and I can check stuff into git - however, the complexity of these source control systems has developed to a point where you can either 'code it' or check-in 'correctly')

so lets just understand that 'dev-ops' is not 'dev' (and vice versa)

Reply:
Don't worry about all the version control talk. That's not a question for now.

I don't think I agree that the second check isn't necessary. The emitter passed to `Observable.create()` is a callback which is called at the point a caller _subscribes_ to the observable. When an observable is first created, it is not subscribed - that happens when someone calls a terminating method on the observable chain - so there could be a lot of time between the `sendMessage` call and this callback getting invoked.

Reply:
let me take a second look, there is a good reason why i felt it was ok to remove it, mainly because it raised the error and 'failed' the subscription request.  It was my view, that you should not need a connection to make a subscription request (as it might be reconnecting while you are requesting - this case I came across).  But i know this is deep in the library, and could have other side affects.

Reply:
Great. Happy to see it go if there's a good reason 👍

Reply:
after looking again, I think subscribe channel should look like this:
`public Observable<T> subscribeChannel(String channelName, Object... args) {
        final String channelId = getSubscriptionUniqueId(channelName, args);
        LOG.info("Subscribing to channel {}", channelId);

        return Observable.<T>create(e -> {
            channels.computeIfAbsent(channelId, cid -> {
                Subscription newSubscription = new Subscription(e, channelName, args);
                try {
                    // sendMessage will log an error if we are not connected, but it does not raise an exception
                    sendMessage(getSubscribeMessage(channelName, args));

                } catch (IOException throwable) { // if getSubscribeMessage throws this, it is because it needs to report
                    e.onError(throwable);    // a problem creating the message, should really be Exception (not IOException)
                }
                return newSubscription;
            });
        }).doOnDispose(() -> {
            if (channels.remove(channelId) != null) {
                try {
                    sendMessage(getUnsubscribeMessage(channelId));
                } catch (IOException e) {
                    LOG.error("Failed to unsubscribe channel: {}",channelId);
                }
            }
    }).share();
}`





As a separate issue, i would like to be able to disable the reconnect logic (so clients can be notified as soon as there is a problem with the socket), by maybe passing a negative retryDuration.


Not sure why this overrides the parameters passed
`
public NettyStreamingService(String apiUrl, int maxFramePayloadLength, Duration connectionTimeout, Duration retryDuration) {
//         this(apiUrl, maxFramePayloadLength, DEFAULT_CONNECTION_TIMEOUT, DEFAULT_RETRY_DURATION, DEFAULT_IDLE_TIMEOUT);

    this(apiUrl, maxFramePayloadLength, connectionTimeout, retryDuration, DEFAULT_IDLE_TIMEOUT)
}
`


` 
   private void scheduleReconnect() {
        if (retryDuration.toMillis() >= 0) {
            LOG.info("Scheduling reconnection");
            webSocketChannel.eventLoop().schedule(
                    () -> connect().subscribe(),
                    retryDuration.toMillis(),
                    TimeUnit.MILLISECONDS);
        } else {
            reconnFailEmitters.forEach(emitter -> emitter.onNext(new Exception("Reconnect disabled")));
        }
    }
`

I might have prefered to add another event, but i am not sure why reconnFailEmitters and connectionSuccessEmitters are lists instead of Subjects (legacy maybe)?